### PR TITLE
THe program with his whole requirments

### PR DESCRIPTION
**Pull request by Omarmanassrah**
This branch contain the whole program the task manages in the FastAPI Framwork it contain creating, updating, deleting and displaying tasks.

### CODE REVIEW

**Review by FaresAbuIram — COMMENTED** [review]
Please remove the Scripts dir and make all required files in one folder with your full name instead of Omargh then omarMan

**Comment by FaresAbuIram** [comment]
I can see that 1902 files have been changed.
pleas remove unnecessary files

**Comment by Omarmanassrah** [comment]
What is the unnecessary files i remove some files

**Comment by Omarmanassrah** [comment]
@FaresAbuIram 
I push right now a file and delete the one that is not needed you can see

File: Omargh/Tasks.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+from typing import Dict, Optional
+from modules import TaskCreate, Task, UrgentTask
+from uuid import UUID, uuid4
+class Task:
+    task_id=0
+    def __init__(self,title,description,due_date,status="InProgress"):
+        Task.task_id+=1
```
Comment (FaresAbuIram) [inline Omargh/Tasks.py:7]: please use the uuid instead 

look here: https://docs.python.org/3/library/uuid.html
use uuid4

File: Omargh/pyvenv.cfg
Comment (FaresAbuIram) [inline Omargh/pyvenv.cfg:1]: we don't need this file

File: Omargh/requirements.txt
Comment (FaresAbuIram) [inline Omargh/requirements.txt:1]: I see nothing here

File: Omargh/Tasks.py (hunk 1 above)
Comment (Omarmanassrah) [reply Omargh/Tasks.py]: Done

File: Omargh/pyvenv.cfg
Comment (Omarmanassrah) [reply Omargh/pyvenv.cfg:1]: okey

File: Omargh/main.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+from modules import Task, UrgentTask,TaskCreate
+from uuid import UUID, uuid4 
+from taskmanager import TaskManager
+
+app = FastAPI()
+# task_manager is made from taskmanager class that work with dictionary
+task_manager = TaskManager()
+""" Creating a Task"""
+@app.post("/tasks/")
+async def create_task(task_data: TaskCreate, is_urgent: bool = False):
```
Comment (FaresAbuIram) [inline Omargh/main.py:12]: I suggest using the priority in the body of the base model 
if the Priority not nil -> urgent task
if Priority nil -> normal task

File: Omargh/main.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+""" Update a Task by ID"""
+@app.put("/tasks/{task_id}")
+def update_tasks(task_id: str , task_data: TaskCreate):
+# using the id to determine the task and taskupdate.dict is refer to pydantic converts the model instance into a dictionary
+    try:
+        updated_task = task_manager.update_task(task_id, task_data)
+        return updated_task
+    except ValueError as e:
+        raise HTTPException(status_code=404, detail="No tasks")
+"""Get All Tasks"""
```
Comment (FaresAbuIram) [inline Omargh/main.py:25]: add space before

File: Omargh/requirements.txt
Comment (Omarmanassrah) [reply Omargh/requirements.txt:1]: may be i delete the enva by mistake should i do new enva and put the files in it ??????

File: Omargh/main.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+async def get_all_tasks():
+    return task_manager.view_tasks()
+"""Get Task by ID"""
+@app.get("/tasks/{task_id}")
+async def get_task(task_id: str):
+    task = task_manager.view_tasksbyid(task_id)
+    if task :
+        return task
+    raise HTTPException(status_code=404, detail="Task not found")
+"""Get Urgent Tasks"""
```
Comment (FaresAbuIram) [inline Omargh/main.py:36]: add space before

File: Omargh/main.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    try:
+        updated_task = task_manager.update_task(task_id, task_data)
+        return updated_task
+    except ValueError as e:
+        raise HTTPException(status_code=404, detail="No tasks")
+"""Get All Tasks"""
+@app.get("/tasks/")
+async def get_all_tasks():
+    return task_manager.view_tasks()
+"""Get Task by ID"""
```
Comment (FaresAbuIram) [inline Omargh/main.py:29]: add tasks before

File: Omargh/main.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+# using the id to determine the task and taskupdate.dict is refer to pydantic converts the model instance into a dictionary
+    try:
+        updated_task = task_manager.update_task(task_id, task_data)
+        return updated_task
+    except ValueError as e:
+        raise HTTPException(status_code=404, detail="No tasks")
+"""Get All Tasks"""
+@app.get("/tasks/")
+async def get_all_tasks():
+    return task_manager.view_tasks()
```
Comment (FaresAbuIram) [inline Omargh/main.py:28]: instead of view_tasks, use get_tasks since the view related to print smth

File: Omargh/modules.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    due_date: date
+    status: str=Field(...,min_length=1,max_length=10)
+# In this case I use separate  class for achieving the singleton Design Pattern Each class have one responsibility
+class TaskCreate(TaskBase):
+    pass
+class Task(TaskBase):
+    task_id:UUID=uuid4()
+
+class UrgentTask(TaskBase):
+     priority: str
```
Comment (FaresAbuIram) [inline Omargh/modules.py:27]: here you can use only on base class model 
`class TaskBase(BaseModel):
    title:str=Field(..., min_length=2, max_length=10)
    description:str=Field(...,min_length=5,max_length=20)
    due_date: date
    status: str=Field(...,min_length=1,max_length=10)
    `
    add  priority to that model and use 
    
    for the uuid4() -> you can move it to the constructor

**Review by FaresAbuIram — CHANGES_REQUESTED** [review]

File: Omargh/modules.py (hunk 7 above)
Comment (Omarmanassrah) [reply Omargh/modules.py]: Also Done

**Comment by Omarmanassrah** [comment]
@FaresAbuIram everything is done if there anything else please tell me

File: my_project_omar/Tasks.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+from uuid import UUID, uuid4
+class Task:
+    
+    def __init__(self,title,description,due_date,status="InProgress"):
+        
+        self.title=title
+        self.description=description
+        self.due_date=due_date
+        self.status=status
+    def update_details(self,title=None, description=None, due_date=None,status=None):
```
Comment (Momen-G-Ar) [inline my_project_omar/Tasks.py:12]: In the `update_details` function: i prefer to accept one object that contains all the variables you need to change. 
After accepting this object: iterate over all its properties using loop _`(You do the same operations many times, do it with loop)`_

File: my_project_omar/taskmanager.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+                updated_task = Task(
+                title=task_data.title,
+                description=task_data.description,
+                due_date=task_data.due_date,
+                status=task_data.status)
+                self.tasks_dict[task_id]=updated_task
+
+    def delete_task(self, task_id: str) -> Optional[Task]:
+        return self.tasks_dict.pop(task_id)
+
```
Comment (Momen-G-Ar) [inline my_project_omar/taskmanager.py:49]: Delete the empty lines

File: my_project_omar/Tasks.py (hunk 8 above)
Comment (Omarmanassrah) [reply my_project_omar/Tasks.py]: Momen I fix this issue but I get some help from the internet then I understand it very well

File: my_project_omar/taskmanager.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+            add_task=Task(title,description,due_date,status)
+            tasks_dict[add_task.task_id]=add_task
+        
+    def get_tasksbyid(self, task_id: str) -> Optional[Task]:
+      return self.tasks_dict.get(task_id)
+    
+    def get_tasks(self)-> Dict[str, Task]:
+      return self.tasks_dict
+#   Default value
+    def update_task(self, task_id: str, task_data: TaskCreate) -> Optional[Task]:
```
Comment (Momen-G-Ar) [inline my_project_omar/taskmanager.py:26]: You edited `update_task` in the `Tasks.py`, you should edit it here too.
Make sure that the user can edit one field at least without being forced to enter all fields. for example: the user can enter one value to edit and must not enter other values, so you edit the value is not none.

**Review by Momen-G-Ar — APPROVED** [review]

File: my_project_omar/modules.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+# Using Enum
+class Status(str, Enum):
+    in_progress = "InProgress"
+    completed = "Completed"
+
+class Priority(str, Enum):
+    high = "High"
+    medium = "Medium"
+    low = "Low"
+class TaskBase(BaseModel):
```
Comment (FaresAbuIram) [inline my_project_omar/modules.py:15]: Where you used this one? 

I suggest to rename it CreateTaskModel and remove the one in line 22

File: my_project_omar/modules.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+class TaskCreate(TaskBase):
+    pass
+class UpdateDetails(BaseModel):
+    title: Optional[str] = None
+    description: Optional[str] = None
+    due_date: Optional[date] = None
+    status: Optional[str] = None
+    priority: Optional[str] = None
+    
+class Task(TaskBase):
```
Comment (FaresAbuIram) [inline my_project_omar/modules.py:31]: Why do you need this  one? 
I think we don't want it at all

File: my_project_omar/modules.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    task_id:UUID=uuid4()
```
Comment (FaresAbuIram) [inline my_project_omar/modules.py:32]: You can move it to the Task class in Tasks.py file

File: my_project_omar/modules.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    def __init__(self, title: str, description: Optional[str], due_date: date, status: str, priority: Optional[Priority] = None, task_id: Optional[UUID] = None):
+        super().__init__(title=title, description=description, due_date=due_date, status=status, priority=priority)
+        self.task_id = task_id 
+class UrgentTask(TaskBase):
```
Comment (FaresAbuIram) [inline my_project_omar/modules.py:36]: The same, use only the class in Task.py 
No need for this one.

File: my_project_omar/main.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+"""Get Task by ID"""
+@app.get("/tasks/{task_id}")
+async def get_task(task_id: str):
+    task = task_manager.get_tasksbyid(task_id)
+    if task :
+        return task
+    raise HTTPException(status_code=404, detail="Task not found")
+
+"""Get Urgent Tasks"""
```
Comment (FaresAbuIram) [inline my_project_omar/main.py:39]: Where is the logic?

File: my_project_omar/modules.py (hunk 11 above)
Comment (Omarmanassrah) [reply my_project_omar/modules.py:15]: Okay

**Review by FaresAbuIram — APPROVED** [review]

**Pull request merged**